Document Maxwell 2's texture headers

Gnurou opened the pull request:
This PR does two things:

1) Update the G80 header documentation to complete them and harmonize the names with NVIDIA's
2) Add the complete documentation for the GM2 texture headers. This new format is the only one recognized by GM200 and later chips.

The renaming of G80 is helpful in that it helps understanding where fields have moved in the new GM2 structures, however it will impact code using the generated headers. If I went too far in the renaming frenzy, please let me know.

Also I am not too sure whether my usage of the inline property in the new enums is correct.

mupuf commented:
Thanks!

imirkin commented on rnndb/g80_defs.xml:
```diff
@@ -72,17 +72,122 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
+	<value value="2" name="UNORM"/>
+	<value value="3" name="SINT"/>
+	<value value="4" name="UINT"/>
+	<value value="5" name="SNORM_FORCE_FP16"/>
```
Can you explain what this name means? Before we had SSCALED, but pretty sure we never used it, so it could well have been wrong.

Gnurou replied:
The doc says "native 8/16bit snorm components forced to FP16 format". I have added that information, although the name already somehow contains it.

imirkin commented on rnndb/graph/g80_texture.xml:
```diff
@@ -6,157 +6,312 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
-			<value value="0x38" name="Z32_X8_C8_X16_MS4_CS12" />
-			<value value="0x39" name="Z32_S8_C8_X16_MS4_CS12" />
-			<value value="0x3a" name="Z16" />
+		<bitfield pos="30" name="PACK_COMPONENTS">
```
Please confirm variants, unless you're saying this was available on G80

Gnurou replied:
This field seems to have appeared in G84.

imirkin commented on rnndb/graph/g80_texture.xml:
```diff
@@ -6,157 +6,312 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
+			<value value="0x17" name="BC7U" variants="GF100-">
+				<doc>similar to ZOL (UNORM8)</doc>
+			</value>
+			<value value="0x06" name="ETC2_RGB"/>
```
variants on all the ETC/EAC formats? guessing GK104+ or GM107+?

imirkin replied:
FWIW after a bit of futzing on GK208 I couldn't get this to work. Certainly an option here is that I messed it up, but I was also seeing this in dmesg:
```
[175799.299477] nouveau 0000:01:00.0: gr: TRAP ch 2 [003fbe1000 oes_compressed_[6736]]
[175799.299484] nouveau 0000:01:00.0: gr: GPC0/TPC0/TEX: 80000009
[175799.299488] nouveau 0000:01:00.0: gr: GPC0/TPC1/TEX: 80000009
```
Guessing it's GM107+?

Gnurou replied:
After some research it seems to have appeared with GK104+. Not sure why you would get this error. :/

imirkin commented on rnndb/graph/g80_texture.xml:
```diff
@@ -6,157 +6,312 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
+			<value value="0x39" name="ZF32_X16V8S8__COV4R12V">
+				<doc>stencil,    4+12 VCAA, 32-bit floating-point depth</doc>
+			</value>
+			<value value="0x3a" name="Z16">
```
variants (G200+? Similar for some of the coverage ones, pretty sure they're not all available on G80)

Gnurou replied:
Not clear - but I cannot find any trace of these before fermi, so let's go with GF100+

imirkin commented on rnndb/graph/g80_texture.xml:
```diff
@@ -6,157 +6,312 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
+		<bitfield pos="30" name="DEPTH_TEXTURE" type="boolean">
+			<doc>Tells texture unit to smear border color from R to RGBA channels</doc>
+		</bitfield>
+		<bitfield pos="31" name="USE_TEXTURE_HEADER_V2" type="boolean">
```
G84+?

Gnurou replied:
Seems so, yes.

imirkin commented on rnndb/graph/g80_texture.xml:
```diff
@@ -6,157 +6,312 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
-		<bitfield high="7" low="4" name="MAX_LEVEL" />
-		<bitfield high="15" low="12" name="MS_MODE" type="g80_msaa_mode">
-			<doc>Like g80/gf100_msaa_mode, dimensions are unaffected.</doc>
+		<bitfield high="3" low="0" name="RES_VIEW_MIN_MIP_LEVEL">
```
I guess these should all be G84+, with G80 getting the color key thing? Technically it's not based on chipset, but in practice... Or you can put the variant on the whole reg and then make a G80-only reg that's just called COLOR_KEY or whatever.

Gnurou replied:
I made two variants of this register depending on the header format version used.

imirkin commented on rnndb/graph/g80_texture.xml:
```diff
@@ -6,157 +6,312 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
+		<bitfield high="15" low="12" name="MULTI_SAMPLE_COUNT" type="g80_msaa_mode">
+			<doc>Like g80/gf100_msaa_mode, dimensions are unaffected. return value for the TEX_HEADER_TEXTURE_TYPE ISA instruction</doc>
+		</bitfield>
+		<bitfield high="27" low="16" name="MIN_LOD_CLAMP">
```
type="ufixed" radix="8"

Please do this in other places where you have fixed point bitfields (use type="fixed" for signed)

Gnurou replied:
Should be fixed throughout the patch.

imirkin commented on rnndb/graph/gm200_texture.xml:
```diff
@@ -0,0 +1,390 @@
+				<doc>stencil,    8+24 VCAA, 32-bit floating-point depth</doc>
+			</value>
+
+			<value value="0x40" name="ASTC_2D_4X4"/>
```
Please confirm that these are all available on GM107

Gnurou replied:
They are. Interestingly they also seem to be available on GK20A but not other Kepler chips.

imirkin commented on rnndb/graph/gm200_texture.xml:
```diff
@@ -0,0 +1,390 @@
+			<value value="0" name="ONE"/>
+			<value value="1" name="TWO"/>
+			<value value="2" name="FOUR"/>
+			<value value="3" name="HEIGHT"/>
```
Maybe that's how it's spelled in french...

Gnurou replied:
Urgh, fixed. :)

imirkin commented on rnndb/graph/gm200_texture.xml:
```diff
@@ -0,0 +1,390 @@
+			<value value="7" name="SAMPLER_COLOR">
+				<doc>border comes from sampler</doc>
+			</value>
+			<doc>border may be either constant color (from sampler), 1, 2, 4, or 8</doc>
```
If you're able to, please elaborate, as this comment makes no sense to me.

Gnurou replied:
Cannot find more information about this, but your explanation is the only one that makes sense.

imirkin commented on rnndb/graph/g80_texture.xml:
```diff
@@ -6,157 +6,312 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
 
 <import file="nvchipsets.xml"/>
 <import file="g80_defs.xml"/>
+<import file="nv_defs.xml"/>
```
why is this necessary?

Gnurou replied:
rnn will complain about inv_memory_layout being an unknown type without this. But this was already the case before this change, so let me move this into its own patch.

koriakin commented on rnndb/graph/g80_texture.xml:
```diff
@@ -6,157 +6,312 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
+			<doc>Tells texture unit to smear border color from R to RGBA channels</doc>
+		</bitfield>
+		<bitfield pos="31" name="USE_TEXTURE_HEADER_V2" type="boolean">
+			<doc>if 0, then byte 7 of the TIC is used to store the color key value. Should always be 1.</doc>
```
you mean word 7, right?

Gnurou replied:
Correct, thanks.

koriakin commented on rnndb/graph/g80_texture.xml:
```diff
@@ -170,45 +325,53 @@ xsi:schemaLocation="http://nouveau.freedesktop.org/ rules-ng.xsd">
+		<bitfield high="30" low="26" name="TRILIN_OPT">
+			<doc>slope for trilinear optimization, e2m3, controlled by USE_HEADER_OPT_CONTROL</doc>
+		</bitfield>
 		<value value="0x10000000" name="UNKN_ANISO_15" />
```
those 2 should be killed off, they refer to TRILIN_OPT bits

Gnurou replied:
Done, thanks.

koriakin commented on rnndb/graph/gm200_texture.xml:
```diff
@@ -0,0 +1,390 @@
+
+<domain name="TIC" size="0x20" prefix="chipset" variants="GM107-">
+	<brief>Texture image control block (for Maxwell)</brief>
+	<doc><p>For Maxwell, the texture headers were substantially redesigned, and five versions are now used. The HEADER_VERSION field of Dword 2 controls which version is used. Many fields change depending on the version, so one has to be extra careful with that.</p><p>GM10X still understands the older format, and uses it by default. It can optionally use the new format if the SetSelectMaxwellTexture ...
```
please add some newlines here

Gnurou replied:
Fixed formatting of other long lines as well.

koriakin commented on rnndb/graph/gm200_texture.xml:
```diff
@@ -0,0 +1,390 @@
+
+	<reg32 offset="0x4" name="1_PITCH" varset="TIC_HEADER_VERSION" variants="PITCH_COLORKEY PITCH">
+		<brief>used if (HEADER_VERSION == PITCH_COLORKEY || HEADER_VERSION == PITCH)</brief>
+		<bitfield high="31" low="5" name="ADDRESS_BITS_31_TO_5">
```
add shr="5" so that values will  be auto-shifted by 5

Gnurou replied:
Done.

koriakin commented on rnndb/graph/gm200_texture.xml:
```diff
@@ -0,0 +1,390 @@
+		<bitfield high="6" low="5" name="GOB_DEPTH_OFFSET">
+			<doc>offset from the base in the depth dimension, to account for sub-allocation of textures with 3D gobs</doc>
+		</bitfield>
+		<bitfield high="31" low="9" name="ADDRESS_BITS_31_TO_9">
```
also shr="9"

Gnurou replied:
Done.

imirkin commented on rnndb/graph/gm200_texture.xml:
```diff
@@ -0,0 +1,390 @@
+	<value value="0x4" name="BLOCKLINEAR_COLORKEY"/>
+</enum>
+
+<domain name="TIC" size="0x20" prefix="chipset" variants="GM107-">
```
I wonder if this should be called TIC2. On GM107 both this and the other one will be available, and I want to be able to include this file in rnndb/root.xml

Gnurou replied:
You're right. Updated.

koriakin commented on rnndb/graph/gm200_texture.xml:
```diff
@@ -0,0 +1,390 @@
+
+	<reg32 offset="0xc" name="3_PITCH" varset="TIC_HEADER_VERSION" variants="PITCH_COLORKEY PITCH">
+		<brief>these fields are available if (HEADER_VERSION == PITCH_COLORKEY || HEADER_VERSION == PITCH)</brief>
+		<bitfield high="15" low="0" name="PITCH_BITS_20_TO_5">
```
shr="5"

Gnurou replied:
Fixed.

koriakin commented on rnndb/graph/gm200_texture.xml:
```diff
@@ -0,0 +1,390 @@
+			<doc>depth of block in gobs</doc>
+		</bitfield>
+		<bitfield high="12" low="10" name="TILE_WIDTH_IN_GOBS" type="g80_gobs_per_block">
+			<doc>width of the sparse tile in gobs</doc>
```
huh, sparse tile?

Gnurou commented:
Pushed v2 addressing all the comments received. Also added a missing field in g80 that only exists on GK20A (FWIW) and merged all register variants for gm200 as suggested by mwk on IRC.
